Tools/TestWebKitAPI/cocoa/TestWKWebView.mm

```diff
@@ -746,34 +746,34 @@ - (void)evaluateJavaScriptAndWaitForInputSessionToChange:(NSString *)script
     }
 }
 
-- (RetainPtr<NSArray>)selectionRectsAfterPresentationUpdate
+- (CGRect)caretViewRectInContentCoordinates
 {
-    RetainPtr<TestWKWebView> retainedSelf = self;
+    UIView *caretView = [self.textInputContentView valueForKeyPath:@"interactionAssistant.selectionView.caretView"];
 
-    __block bool isDone = false;
-    __block RetainPtr<NSArray> selectionRects;
-    [self _doAfterNextPresentationUpdate:^() {
-        selectionRects = [retainedSelf _uiTextSelectionRects];
-        isDone = true;
-    }];
-
-    TestWebKitAPI::Util::run(&isDone);
-    return selectionRects;
-}
+#if HAVE(UI_TEXT_SELECTION_DISPLAY_INTERACTION)
+    if (!caretView) {
+        if (auto view = self.textSelectionDisplayInteraction.cursorView; !view.hidden)
+            caretView = view;
+    }
+#endif
 
-- (CGRect)caretViewRectInContentCoordinates
-{
-    UIView *selectionView = [self.textInputContentView valueForKeyPath:@"interactionAssistant.selectionView"];
-    CGRect caretFrame = [[selectionView valueForKeyPath:@"caretView.frame"] CGRectValue];
-    return [selectionView convertRect:caretFrame toView:self.textInputContentView];
+    return [caretView convertRect:caretView.bounds toView:self.textInputContentView];
 }
 
 - (NSArray<NSValue *> *)selectionViewRectsInContentCoordinates
 {
     NSMutableArray *selectionRects = [NSMutableArray array];
-    NSArray<UITextSelectionRect *> *rects = [self.textInputContentView valueForKeyPath:@"interactionAssistant.selectionView.rangeView.rects"];
-    for (UITextSelectionRect *rect in rects)
-        [selectionRects addObject:[NSValue valueWithCGRect:rect.rect]];
+    NSArray *rects = [self.textInputContentView valueForKeyPath:@"interactionAssistant.selectionView.rangeView.rects"];
```

Comment:
Can't we keep the type as `NSArray<UITextSelectionRect *>`, and use `UITextSelectionRect` (and the `rect` property) in the loop below?

Reply:
Yes — good catch.